1. Create NPV and IRR calculator 
2. Create User Interface (functions) to get the required data from the user and complete NPV and IRR analysis
3. Conduct calculator and user input tests

Student Name: Collin Czernel

Student ID: 

We are going to create a NPV and IRR calculator function. We will then create an interface to allow user inputs for cash flow, principal amount (at t = 0), rate of return, and time duration.

NPV stands for Net Present Value, which is a method of seeing if an investment will make or lose money based on the "time value of money" (e.g. 100 dollars today is worth more than 100 dollars in a year). The initial cash flow is usually negative (as its the investment amount) and the following cash flows are the returns expected each year. We apply a discount to the money earned in the future to reflect what its worth today, and that is all added up. If the NPV is positive, one would expect the venture to be profitable, and if its negative, one would expect to lose money. 

IRR stands for Internal Rate of Return, which is a percentage that "shows" how profitable an investment is expected to be. The IRR is the discount rate (applied in the paragraph above) at which the NPV = 0. If the IRR is higher than a typical investment savings at 3%, it would be considered a "good" investment.

In [19]:
import numpy_financial as npf

In [35]:
def npv_calculator(ror, principal, cash_flows): # where "ror" is the rate of return, principal is the initial amount, cash_flow is the cash flow list)
    npv = -principal # The first investment is the negative to balance the potential future earnings as the principal investment 
    for t, cash in enumerate(cash_flows): # loop through the npv and sum up the values from the formula
        npv += cash/(1+ror) ** (t+1)
    return npv 

def irr_calculator(principal, cash_flows):
    cash_flow_list = [-principal] + cash_flows # negative for initial investment, creating list with principal amount (as negative) and adding cash flows
    irr = npf.irr(cash_flow_list) # using numpy financial module to do the IRR calc on the cash flow list
    return irr

def evaluate_project(npv, irr, discount):
    if npv > 0 and irr >= discount:
        print("PROJECT RECOMMENDED")
    elif npv < 0 or irr < discount:
        print("PROJECT NOT RECOMMENDED")
    elif npv == 0 and irr >= discount:
        print("No gain or loss - investment is pointless")

In [39]:
principal = int(input("Enter the initial investment:"))
cash_flows = []

period = int(input("Enter the time frame (in years):")) # We want to use an int because we're only working with whole years
for i in range(period): # depending on how many years t is 
    cash_flow = float(input(f"Enter the cash flow for period {i + 1}: ")) # float for non-whole ints
    cash_flows.append(cash_flow) # add to cash flows list

discount = float(input("Enter the discount rate (in decimal):")) # have to add decimal here for the discount rate, usually using 3% in examples
npv_result = npv_calculator(discount, principal, cash_flows) # find the npv of our input values
irr_result = irr_calculator(principal, cash_flows) # find the irr of our input values

print("The NPV is: ", npv_result) # print results
print("The IRR is: ", irr_result) # print results
evaluate_project(npv_result, irr_result, discount)

Enter the initial investment: 100
Enter the time frame (in years): 3
Enter the cash flow for period 1:  25
Enter the cash flow for period 2:  25
Enter the cash flow for period 3:  25
Enter the discount rate (in decimal): 0.10


The NPV is:  -37.828700225394456
The IRR is:  -0.13112314790418045
PROJECT NOT RECOMMENDED


For a test case where we put in 100, then get 25 back for 3 years, we expect a negative NPV (which we achieve).

In [43]:
# Now testing the examples found in the Excel workbook with error handling 
cash_flows = []
while True: # error handling loop for only accepting floats
    try:
        principal = float(input("Enter the initial investment:")) # floats to allow for non-whole initial investments.
        break
    except ValueError:
        print("The principal amount has to be a number.")

while True: # error handling loop for only accepting ints
    try:
        period = int(input("Enter the time frame (in years):")) # We want to use an int because we're only working with whole years
        break
    except ValueError:
        print("The number of years has to be a whole number.")

for i in range(period): # depending on how many years t is 
    cash_flow = float(input(f"Enter the cash flow for period {i + 1}: ")) # float for non-whole ints
    cash_flows.append(cash_flow) # add to cash flows list

while True: # error handling loop for only accepting floats
    try:
        discount = float(input("Enter the discount rate (in decimal):")) # have to add decimal here for the discount rate, usually using 3% in examples
        break
    except ValueError:
        print("The discount rate should be in decimal form.")
        
npv_result = npv_calculator(discount, principal, cash_flows) # find the npv of our input values
irr_result = (irr_calculator(principal, cash_flows))*100 # find the irr of our input values

print("The NPV is: ", npv_result) # print results
print("The IRR is: ", irr_result, "%") # print results
evaluate_project(npv_result, irr_result, discount)

Enter the initial investment: 750
Enter the time frame (in years): 4
Enter the cash flow for period 1:  210
Enter the cash flow for period 2:  250
Enter the cash flow for period 3:  300
Enter the cash flow for period 4:  250
Enter the discount rate (in decimal): 0.175


The NPV is:  -74.11294918663748
The IRR is:  12.559483650240132 %
PROJECT NOT RECOMMENDED


So, this matches the excel example with an NPV of -74.11 and an IRR of 12.56%. This would then be considered a BAD investment as we are overall losing money, as if we invested the 750 originally in a savings account or a similar-risk venture we would likely do better overall. To further add to this, it would be optimal to put the input into a function to be easily called for future use cases.